sdk/opendp/smartnoise/_ast/expressions/logical.py

```diff
@@ -29,7 +32,56 @@ def children(self):
     def evaluate(self, bindings):
         l = self.left.evaluate(bindings)
         r = self.right.evaluate(bindings)
-        return ops[self.op.lower()](l, r)
+        if (type(l) != type(r)):
+            if isinstance(l, str):
+                if isinstance(r, bool):
+                    l = parse_bool(l)
+                elif isinstance(r, int):
+                    try:
+                        l = int(l)
+                    except:
+                        l = float(l)
+                elif isinstance(r, float):
+                    l = float(l)
+                elif isinstance(r, datetime):
+                    l = datetime.fromisoformat(l)
+                elif isinstance(r, date):
+                    l = date.fromisoformat(l)
```

Comment:
The nested ifs should be a function.

if isinstance(l, str):
   l = resolve_string_type(l, r)
elif isinstance(r, str):
   r = resolve_str_type(r, l)

Reply:
Good call; I have updated